File: 2.py
class Rectangle:
    
    def __init__(self, width, height):
        self.x = 0
        self.y = 0
        self.width = width
        self.height = height
        
    def __del__(self):
        print(f"Деструктор для объекта {self}")
        
    def get_area(self):
        return self.width * self.height
    
    def left(self):
        return self.x
    
    def right(self):
        return self.x + self.width
    
    def bottom(self):
        return self.y
    
    def top(self):
        return self.y + self.height
    
    def __mul__(self, other):
        x1 = max(self.left(), other.left())
        y1 = max(self.bottom(), other.bottom())
        x2 = min(self.right(), other.right())
        y2 = min(self.top(), other.top())
        
        if x2 <= x1 or y2 <= y1:
            return Rectangle(0, 0)
        
        return Rectangle(x2 - x1, y2 - y1)
    
    def __gt__(self, other):
        return self.get_area() > other.get_area()
    
    def __lt__(self, other):
        return self.get_area() < other.get_area()
    
    def __str__(self):
        return f"Rectangle({self.left()},{self.bottom()},{self.width},{self.height})"
        
r1 = Rectangle(3, 4)
r2 = Rectangle(2, 5)

print("Прямоугольник 1: (3, 4)")
print("Прямоугольник 2: (2, 5)")

intersection = r1 * r2

print(f"Пересечение: {intersection}")

print(f"Площадь прямоугольника 1: {r1.get_area()}")
print(f"Площадь прямоугольника 2: {r2.get_area()}")

if r1 > r2:
    print("Площадь прямоугольника 1 больше")
else:
    print("Площадь прямоугольника 2 больше")

File: 2.cpp
#include <iostream>

class Rectangle {
private:
    double x = 0;
    double y = 0;
    double width = 1;
    double height = 1;

public:
    friend std::ostream& operator<<(std::ostream& out, const Rectangle& r);

    friend std::istream& operator>>(std::istream& in, Rectangle& r);

    Rectangle(double width, double height) : width(width), height(height) {}

    ~Rectangle() {
        std::cout << "Деструктор для объекта " << *this << std::endl;
    }

    double getArea() const {
        return width * height;
    }

    double left() const { return x; }
    double right() const { return x + width; }
    double bottom() const { return y; }
    double top() const { return y + height; }

    Rectangle operator*(const Rectangle& other) const {
        double x1 = std::max(left(), other.left());
        double y1 = std::max(bottom(), other.bottom());
        double x2 = std::min(right(), other.right());
        double y2 = std::min(top(), other.top());

        if (x2 <= x1 || y2 <= y1) {
            return Rectangle(0, 0); // нет персечения
        }

        return Rectangle(x2 - x1, y2 - y1);
    }

    bool operator>(const Rectangle & other) const {
        return getArea() > other.getArea();
    }

    bool operator<(const Rectangle& other) const {
        return getArea() < other.getArea();
    }
};

int main(void) {
    Rectangle r1(3, 4);
    Rectangle r2(2, 5);
    std::cout << "rectangle 1 (3,4)" << std::endl;
    std::cout << "rectangle 2 (2,5)" << std::endl;

    auto intersection = r1 * r2;
    std::cout << "Peresechenie: " << intersection << std::endl;

    std::cout << "rectangle 1 area: " << r1.getArea() << std::endl;
    std::cout << "rectangle 2 area: " << r2.getArea() << std::endl;

    if (r1 > r2) {
        std::cout << "rectangle 1 area > rectangle 2 area" << std::endl;
    }
    else {
        std::cout << "rectangle 1 area < rectangle 2 area" << std::endl;
    }
}


std::ostream& operator<<(std::ostream& out, const Rectangle& r) {
    out << "Rectangle(" << r.left() << "," << r.bottom() << ","
        << r.width << "," << r.height << ")";
    return out;
}

std::istream& operator>>(std::istream& in, Rectangle& r) {
    in >> r.x >> r.y >> r.width >> r.height;
    return in;
}